### ISSUE TITLE
COCO evaluation results are zero while training

### ISSUE THREAD

**Issue by amardeepjaiman** [fizyr/keras-retinanet#881]
Hi, I am created my custom dataset in COCO format and trying to train the Retinanet using Resnet50 using default pretrained weights provided. Issue is while training COCO evaluation always giving mAP=0 while the loss is getting decreased in each epoch.

Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.0

Can someone suggest what could be the problem?

Thanks,
Amardeep

**Comment by hgaiser** [comment]
What @martinzlocha said is a good experiment to perform. Also, what are your loss values during training?

**Comment by martinzlocha** [comment]
@amardeepjaiman What was the issue and how did you fix it? This might be very useful for someone else who stumbles upon this question in the future.

**Comment by amardeepjaiman** [comment]
Actually in my coco json file, there were repeated entries for annotations. Thats why evaluation was not correct.

**Comment by sumeetssaurav** [comment]
Hii, Just trying to train retinanet model from scratch on COCO. Could you just let me know the time it  takes to complete one epoch.

Thank You

**Comment by kunjmehta** [comment]
@amardeepjaiman facing the same problem. Can you please elaborate on what repeated entries were you getting? Was it the file name or the segmentation/annotation data? And did you just delete the duplicates?

**Comment by WalterYeYint** [comment]
I'm facing the same issue right now using ssd_mobilenet_v2_coco_2018_03_29. @amardeepjaiman , could you please elaborate on how you solved the problem?
I have been training fine using the same .ipynb script until last week when the training process starts outputting mAP and AR of zero values (my guess is due to the new version of python-pil)
Any other helps are appreciated.

**Comment by YasmineAitMimoun** [comment]
Hello, 
I had the same mistake. The problem in my case was to set the label of the only class I had to 0 instead of 1. Indeed 0 must be the label of the background.

**Comment by VtojeC** [comment]
@YasmineAitMimoun 
It didn't work in my case. Single category, changed category_id from 0 to 1 with no results in both DT json and Annonation json files.

**Comment by VtojeC** [comment]
After several days of research I gathered all data that helped me in my evaluation of custom dataset on COCO API:

**Following tips are only valid for evaluation of detection for a single categorie using Bounding Boxes!**

**1. Each annotation in the annotation file MUST have an unique ID.**
I got tripped on this one and I hope nobody will in the future. If annotation ID repeats then calculations get worse and lead to 0.000. This is why you need to check your annotation file if your evaluation doesn't work @kunjmehta like @amardeepjaiman and I did. It really helped. I tested it too.

**2. If you have only a SINGLE CATEGORY in your dataset then your Annotation file and Result file MUST have that category saved as a category != 0.** 
 It can be 1, but DO NOT save your only category as 0, because your evaluation will print only ‘-1.000’.
@YasmineAitMimoun was fully right on this one, thank you. You helped me too. I tested it and it's confirmed.

**3. AREA in detection with bounding boxes & evaluation is basically WIDTH * HEIGHT.**
A lot of articles on the Internet explained it too, but gave bad examples that made me overthink and research more about. Let's make it clear now.

**4. SEGMENTATION in detection with bounding boxes & evaluation is NOT NEEDED.** 
Segmentation in evaluation of detection with bounding boxes in the Annotation file is just an empty list and doesn't have to be filled. I tested it and no matter what you put inside - if you do the same evaluation as I did those values will never be asked for.
Segmentation is a list of verticies of the field that covers detected object/objects, but in bounding boxes it's completely not needed.

**5. Annonation file must contain a list of all images, but each image must be saved only ONCE.**
If an image is mentioned several times then the evaluation won't work, giving the "data doesn’t correspond" message.

**6. Annotation file must contain the list of all images (or more) where detection was conducted on.** 
If your Results file has result for an image that does not simply exists in the Annotation file your evaluation won't work, giving you the same error: "data doesn’t correspond".

I made several mistakes when evaluating COCO, but I would be an ignorant if I didn't mention that official COCO website lacks of good explanation what certain things are meant for. I would be grateful if they corrected and added some more stuff. I wrote everything I tripped on with hope that nobody will go through that hell again.

Good luck.

Edit:
I worked with Windows 11 in PyCharm, using PythonAPI.
